### PR TITLE
feat(web): Remove from Stack

### PR DESCRIPTION
**Pull request by xCJPECKOVERx**
## Description

Adds a menu option for stack assets to remove them from the stack. Only available on stacks with more than 2 assets, and not on the primary asset.
Works by calling stack.create, and adds a 'merge' field to the StackCreateDto. When removing from the stack is does not merge in the assets of existing stack children. Ensures that when adding to or creating a stack, merge is true.
When the asset is removed, the asset-viewer's asset is set to the stack's primary asset.

Updated the icon for Unstack, so that remove can use the image-minus icon.

## How Has This Been Tested?

Create a stack, choose a stack asset that's not the primary, click 'remove from stack' in the navbar menu. The asset-viewer stack will be updated, and the asset-viewer will display the stack's primary asset. Closing the viewer will show the stack count updated, and the removed asset visible in the timeline.

<details><summary><h2>Screenshots (if appropriate)</h2></summary>

image

</details>

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if applicable
- [x] I have no unrelated changes in the PR.
- [x] I have confirmed that any new dependencies are strictly necessary.
- [x] I have written tests for new code (if applicable)
- [x] I have followed naming conventions/patterns in the surrounding code
- [x] All code in `src/services/` uses repositories implementations for database calls, filesystem operations, etc.
- [x] All code in `src/repositories/` is pretty basic/simple and does not have any immich specific logic (that belongs in `src/services/`)

### CODE REVIEW

**Review by danieldietzler — COMMENTED** [review]
I feel like an endpoint like `DELETE stacks/:id/asset/:id` would be better? @jrasm91 thoughts?

**Comment by xCJPECKOVERx** [comment]
Maybe that or the asset's update endpoint to clear the stack id? I used create since that's how updates to the stack (adding to it) are handled already: delete the existing stacks and create a new one with the required assets.

**Comment by danieldietzler** [comment]
Yeah I saw that. I just don't really like the ergonomics of having to create a new stack. That's similar to extending the update endpoint, there we'd also need to enumerate all stack assets again which seems kind of annoying to deal with. That's why I was suggesting such a DELETE endpoint

**Comment by jrasm91** [comment]
Yeah let's have remove an asset from a stack be a separate endpoint. I like this one

```
DELETE /stacks/:stackId/assets/:asset id
```

The create endpoint deals with creating a new stack from a combination of existing stacks and assets, which is why it is a bit more complicated. We don't need any such logic to remove a single asset from a single stack. 

I think the only edge case we should make sure to test is what to do when the second to last asset is removed from the stack, since that implies there is no longer a stack. We should probably throw a BadRequest exception in the case since they should have just used delete stack instead

**Comment by xCJPECKOVERx** [comment]
Sounds good, I can work on it over the next week or so. 
For the second-to-last asset issue, I agree with an error. On front end my plan is to not allow removal if there's only 2 assets anyway.

Another edge case is removing the primary asset, if allowed. I'm currently not allowing it, though arguably wouldn't be too hard to handle in a new endpoint if it can optionally return the new primary asset for the front end to display.

**Comment by danieldietzler** [comment]
You could even just have the frontend re-fetch the stack anyways after deleting that asset

**Comment by xCJPECKOVERx** [comment]
That would depend on whether we want to keep viewing the stack (update the viewing asset), or stay on the old primary which is no longer a part of the stack.

**Comment by danieldietzler** [comment]
Yeah that's true. Anyways, it's not super important right now and can easily be added later on :)

**Comment by jrasm91** [comment]
I think we could just throw an error in that case. We do have the `updateStack` method, which let's you change the primary asset id to something else. The client could have the same "remove from stack option" for all assets, and optionally change the primary if needed before removing it from the stack.

**Comment by xCJPECKOVERx** [comment]
Sounds good.
So return error if:
1. Asset isn't in the stack
2. Asset is the stack's primary
3. There's only 2 assets in the stack


I'll add the endpoint and revert the updates to stackRepo.create and the StackCreatDto.

File: web/src/lib/components/photos-page/asset-grid.svelte [hunk 1]
```diff
@@ -523,6 +523,23 @@
...
+          updateUnstackedAssetInTimeline(
+            timelineManager,
+            action.stack.assets.map((asset) => toTimelineAsset(asset)),
+          );
+          updateStackedAssetInTimeline(timelineManager, {
+            stack: action.stack,
+            toDeleteIds: action.stack.assets
+              .filter((asset) => asset.id !== action.stack?.primaryAssetId)
+              .map((asset) => asset.id),
+          });
```
Comment (jrasm91) [inline web/src/lib/components/photos-page/asset-grid.svelte:539]: We should probably just add a method to the timeline manager to more efficiently handle updating the stack count in the timeline, instead of doing this little dance.

File: server/src/services/stack.service.ts [hunk 2]
```diff
@@ -58,6 +58,40 @@ export class StackService extends BaseService {
     await this.eventRepository.emit('StackDeleteAll', { stackIds: dto.ids, userId: auth.user.id });
   }
 
+  async removeAsset(auth: AuthDto, stackId: string, assetId: string): Promise<StackResponseDto> {
+    await this.requireAccess({ auth, permission: Permission.STACK_UPDATE, ids: [stackId] });
+    await this.requireAccess({ auth, permission: Permission.ASSET_UPDATE, ids: [assetId] });
```
Comment (jrasm91) [inline server/src/services/stack.service.ts]: I think it's fine to skip this check. Stack update is probably sufficient for this.

File: server/src/services/stack.service.ts [hunk 3]
```diff
@@ -58,6 +58,40 @@ export class StackService extends BaseService {
...
+    if (stack.assets?.length <= 2) {
+      throw new BadRequestException('Cannot remove last or second to last asset');
+    }
+
+    //Verify the asset is not the stack's primary asset
+    if (stack.primaryAssetId === assetId) {
+      throw new BadRequestException("Cannot remove stack's primary asset");
+    }
+
+    await this.assetRepository.update({ id: assetId, stackId: null });
```
Comment (jrasm91) [inline server/src/services/stack.service.ts]: Doing a bunch of these database queries (especially in sequence) isn't ideal. Maybe we can try to combine some of the selects to get all the information we need. We could do something like "select asset.stackId, stack.primaryAssetId from assets left join stacks on asset.stackId = stack.id  where asset.id = ?"

Thinking about it more, I'm not sure if we need to do the <= 2 assets check either. Today you can have a stack with 2 assets and delete an asset and the system doesn't clean up the stack with one asset (as far as I can tell), so I'd be fine with removing that check.

File: server/src/controllers/stack.controller.ts [hunk 4]
```diff
@@ -54,4 +66,14 @@ export class StackController {
   deleteStack(@Auth() auth: AuthDto, @Param() { id }: UUIDParamDto): Promise<void> {
     return this.service.delete(auth, id);
   }
+
+  @Delete(':id/assets/:assetId')
+  @Authenticated({ permission: Permission.STACK_UPDATE })
+  removeAssetFromStack(
+    @Auth() auth: AuthDto,
+    @Param() { id }: UUIDParamDto,
+    @Param('assetId', new ParseUUIDPipe()) assetId: string,
```
Comment (jrasm91) [inline server/src/controllers/stack.controller.ts]: Let's just make another dto (similar to UUIDParamDto) that has `id: string, assetId: string` with `@ValidateUUID()` on both and use that here instead.

File: server/src/services/stack.service.ts [hunk 5]
```diff
@@ -58,6 +58,40 @@ export class StackService extends BaseService {
...
+    await this.eventRepository.emit('StackUpdate', { stackId, userId: auth.user.id });
+
+    const updatedStack = {
+      ...stack,
+      assets: stack.assets?.filter((a) => a.id !== assetId),
+    };
+    return mapStack(updatedStack, { auth });
```
Comment (jrasm91) [inline server/src/services/stack.service.ts]: I don't think it's necessary to re-select the entire resource only to send it back without the item that was deleted. What do you think about just changing this delete endpoint to return 204, no content and remove this.

**Comment by jrasm91** [comment]
Also, sorry about the recent merge conflicts :sweat: , I've been renaming/standardizing a bunch of stuff. I'm happy to resolve them if you want, after you've made the other changes.

File: web/src/lib/components/photos-page/asset-grid.svelte (hunk 1 above)
Comment (xCJPECKOVERx) [reply web/src/lib/components/photos-page/asset-grid.svelte:539]: I could take a look at this in a separate PR. No super knowledgeable in how the timeline manager works quite yet.

File: server/src/services/stack.service.ts (hunk 3 above)
Comment (xCJPECKOVERx) [reply server/src/services/stack.service.ts]: > combine some of the selects

Just to confirm correct procedure: for helper queries like this, I should just add a repository function to return a custom (or existing dto) object? 
For this particular case though, I could remove the asset fetch and just get the stack and verify the asset's in there. If the asset doesn't exist, it wouldn't be in the stack anyway.

> I'm not sure if we need to do the <= 2 assets check

If we don't require it, then there should be some stack cleanup. imo it's yucky having an asset showing as a stack of 1. Alternatively, we could remove the check here, but still hide the menu option if there's only 2 in the stack.

File: server/src/services/stack.service.ts (hunk 5 above)
Comment (xCJPECKOVERx) [reply server/src/services/stack.service.ts]: Yup, could do that and then let the client handle updating/filtering the stack assets?

**Comment by xCJPECKOVERx** [comment]
> Also, sorry about the recent merge conflicts 😓 , I've been renaming/standardizing a bunch of stuff. I'm happy to resolve them if you want, after you've made the other changes.

No worries! I noticed some of them, shouldn't be too bad merging the rest once I'm done with the final updates

File: server/src/services/stack.service.ts (hunk 3 above)
Comment (jrasm91) [reply server/src/services/stack.service.ts]: For the helper query, just add a new repository method on the `StackRepository` and you don't need an explicit return type. You can see how the other methods work, especially the asset job repository. You can dynamically select the data you need. I'm also happy to do this for you if you don't want to bother with it.

My point was mainly that we don't need a server-side check here since there are still other ways the user can get into that state. I agree it would be good to add a nightly job or similar that takes care of cleaning up this type of data.

File: server/src/services/stack.service.ts (hunk 5 above)
Comment (jrasm91) [reply server/src/services/stack.service.ts]: Yeah, it should have enough information client-side to figure out what needs to be updated.

File: server/src/services/stack.service.ts (hunk 3 above)
Comment (xCJPECKOVERx) [reply server/src/services/stack.service.ts]: Gotcha. Currently reworking it a bit and I can omit the asset fetch entirely. With that being the case, is a new function really necessary/performative enough over just getting the stack with the assets (like in update's primaryAssetId check)?

Comment (jrasm91) [reply server/src/services/stack.service.ts]: There are two strategies I have seen.
1. Have a method that loads the entire object and use that method everywhere
2. Have specialized methods that load minimal data and use each one in one place.

We used to use TypeORM and option (1) everywhere. The result was a lot of overhead and inefficiency. We've been transitioning to option (2) instead. A lot of our jobs used to load a lot of data that they never used. Those types of queries have been replaced with specialized/minimal ones, which you can see in the `AssetJobRepository`.

I think it's more efficient, faster, and less memory intensive to use a specialized query and only load the info you need in order to do the required checks in this method. For me, I see no reason to load every asset in the stack, which technically could include like 1,000+ assets (since I don't believe we have any limitation on this). Especially if all you need is two booleans (is the asset in the stack, is the asset the stack primary)

BTW - If you want to ping me on discord in the contributing channel and get quicker answers, that's fine with me (`@jrasm91`)

File: web/src/lib/components/photos-page/asset-grid.svelte (hunk 1 above)
Comment (jrasm91) [reply web/src/lib/components/photos-page/asset-grid.svelte:539]: OK. You can leave it as is and I'll maybe take a look before merging the PR.

File: server/src/services/stack.service.ts (hunk 3 above)
Comment (xCJPECKOVERx) [reply server/src/services/stack.service.ts]: Cool, yeah I'll add a specialized fetch then.

> ping me on discord

will do lol!

File: server/src/validation.ts [hunk 6]
```diff
@@ -70,6 +70,18 @@ export class UUIDParamDto {
   id!: string;
 }
 
+export class UUIDAssetIDParamDto {
+  @IsNotEmpty()
+  @IsUUID('4')
+  @ApiProperty({ format: 'uuid' })
```
Comment (jrasm91) [inline server/src/validation.ts]: Should be using `@ValidateUuid()`

File: server/src/services/stack.service.ts [hunk 7]
```diff
@@ -58,6 +59,26 @@ export class StackService extends BaseService {
     await this.eventRepository.emit('StackDeleteAll', { stackIds: dto.ids, userId: auth.user.id });
   }
 
+  async removeAsset(auth: AuthDto, dto: UUIDAssetIDParamDto): Promise<void> {
+    const { id: stackId, assetId } = dto;
+    await this.requireAccess({ auth, permission: Permission.StackUpdate, ids: [stackId] });
+
+    const stack = await this.stackRepository.getForAssetRemoval(assetId);
+
+    //Verify the asset exists and is in the stack
```
Comment (jrasm91) [inline server/src/services/stack.service.ts]: We can remove these two comments

**Review by jrasm91 — APPROVED** [review]
